### PR TITLE
Cache `charmcraft pack` container in integration tests

### LINKED ISSUE #75: Cache `charmcraft pack` container in integration tests
Cache LXC container used by `charmcraft pack` in GitHub Actions integration tests (to speed up tests)

### PR DESCRIPTION
**Pull request by carlcsaposs-canonical**
# Issue
CI takes a long time to run

# Solution
Cache LXC container used by `charmcraft pack` in GitHub Actions integration tests (to speed up tests)

Closes #75 

# Context
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows

# Testing
Tested on https://github.com/carlcsaposs-canonical/mysql-operator

# Release Notes
Cache LXC container used by `charmcraft pack` in GitHub Actions integration tests (to speed up tests)

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -10,159 +16,270 @@ jobs:
...
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Install dependencies
-        run: python -m pip install tox
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
+        run: python3 -m pip install tox
       - name: Run tests
-        run: tox -e unit
+        run: tox run -e unit
```
Comment (taurus-forever) [inline .github/workflows/ci.yaml:35]: It changes nothing for sure, JFMI, why are you adding it? We use "tox -e unit" in all other repos.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -10,159 +16,270 @@ jobs:
...
+        run: |
+          # Copied from https://github.com/charmed-kubernetes/actions-operator/blob/96fb0b07eb675f74cf1796be812bc7e67a0d62fc/src/bootstrap/index.ts#L151
+          sudo adduser $USER lxd
+          newgrp lxd
+          sudo lxd waitready
+          sudo lxd init --auto
+          sudo iptables -F FORWARD
+          sudo iptables -P FORWARD ACCEPT
+          sudo chmod a+wr /var/snap/lxd/common/lxd/unix.socket
+          sudo snap install charmcraft --classic
```
Comment (taurus-forever) [inline .github/workflows/ci.yaml:74]: Did you extract this for the speed reason only? Or smth conceptually broken inside charmed-kubernetes/actions-operator@main?

**Review by taurus-forever — COMMENTED** [review]
1) Wow! I like it! (c)
2) Should we consider to extract matrix / caching logic outside of ci.yaml ? It becomes a monster. Sure a separate topic/PR... maybe new GH Action? CI/CD readability is important.
3) some question inside the review.
4) Included Jarred for CI/CD comments/ideas.

**Comment by taurus-forever** [comment]
> > ```
> > 2. Should we consider to extract matrix / caching logic outside of ci.yaml ? It becomes a monster. Sure a separate topic/PR... maybe new GH Action? CI/CD readability is important.
> > ```
> 
> I experimented with using a reusable workflow and decided to not go that route because (1, primarily) it adds a level of nesting to the UI output of GitHub Actions and makes the logs harder to read/understand and (2) it would require constraining the CI with inputs (e.g. assuming & enforcing a `lint`, `unit-test`, and `integration-` tox environments) and I especially didn't want to constrain the CI when we're still figuring out how to stabilize/mark flaky tests, potentially pin dependencies, and potentially run the tests in smaller modules/more concurrently to speed up CI.

I see your arguments. What about to save python code "name: Prune cache from previous runs" into some script and call it from ci.yaml. It should change nothing on GH UI. Also will allow us to run all python static tests on the script and make ci.yaml better readable.

**Comment by shayancanonical** [comment]
@carlcsaposs-canonical i like  the prune cache logic in a separate file - it makes it more readable and maintainable. if alex agrees, would you mind pulling those changes into this branch (so that we can review them together with the other changes in this PR)?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -1,6 +1,12 @@
+# Copyright 2022 Canonical Ltd.
```
Comment (taurus-forever) [inline .github/workflows/ci.yaml:1]: 2023? :-D

**Review by jardon — COMMENTED** [review]
This is a beast of a PR.  I like whats going on here, but I'd like to think theres some simplicity that can be had here.  I have a couple of questions:

1) Is there any way we can utilize something like yq for creating the matrix instead of having to maintain python code for that?

2) Could we split up the workflows that *dont* need the matrix?  

As is, reusability and maintainability are going to suffer.

**Comment by shayancanonical** [comment]
I agree with @jardon. Some feedback:
1. After tracing the code, I feel like I understand the high level picture, but it is hard to verify the small details of why something is done in `ci.yaml` or the prune_cache script. If we can document the flow and some decisions made, i think it would make this more maintainable
2. I wonder if there's a cleaner, simpler way to abstract the caching of lxc containers (since this will need to be adopted across most of our charm repos)

Basically, I'm thinking of how 1) new members to the feature can easily understand what's going on and 2) how to easily port this to other repos

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -1,6 +1,12 @@
...
+# See LICENSE file for licensing details.
 name: Tests
+
 on:
   pull_request:
+  push:
+    branches:
+      - main
```
Comment (shayancanonical) [inline .github/workflows/ci.yaml:9]: i remember these tests intentionally not being run on push to the main branch. this is because the tests are run [here](https://github.com/canonical/mysql-operator/blob/main/.github/workflows/release.yaml#L23-L26) in `release.yaml`

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -10,159 +14,160 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Setup operator environment
-        uses: charmed-kubernetes/actions-operator@main
+      - name: Setup environment
+        # TODO: Replace with custom image on self-hosted runner
+        run: |
+          # Copied from https://github.com/charmed-kubernetes/actions-operator/blob/96fb0b07eb675f74cf1796be812bc7e67a0d62fc/src/bootstrap/index.ts#L151
```
Comment (shayancanonical) [inline .github/workflows/ci.yaml:64]: suggestion: what if all the bash command sections in this file are broken into a directory with bash files (and these files are invoked in the appropriate step)? 

this allows us to create easily readable and easily portable sections of the workflow. additionally, it provides an opportunity to pepper the bash scripts with comments

File: .github/workflows/prune-cache.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    logging.info(
+        f"{bytes_available / GIBIBYTE:.1f} GiB available, {bytes_required / GIBIBYTE:.1f} GiB required"
+    )
+    if not cache_deletion_order:
+        break
+    # Delete current ref cache if it exists or delete the oldest cache
+    ref, cache = cache_deletion_order.pop(-1)
+    if ref == "refs/heads/main":
+        continue
+    delete_cache(cache)
```
Comment (shayancanonical) [inline .github/workflows/prune-cache.py:130]: thoughts on adding a markdown file somewhere that details how the caching/pruning works? the goal should be allow any new comer to the feature to easily pinpoint what the code does and make any changes necessary

this file is especially daunting for me to trace in detail

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (taurus-forever) [reply .github/workflows/ci.yaml:1]: Normally it should represent the current year. No need to change it here (now). I will check official Canonical policy first.

**Comment by taurus-forever** [comment]
The "Goal" is a fast tests. "Caching" is only a solution.

Demo explained this PR a lot. Tnx @carlcsaposs-canonical !
Still, "+1" to `CI/CD readability decreased`. (c) @jardon 

We are caching to:
  1) speedup charm building
  2) reuse cached charms for tests

PR should be way smaller IF we:
 1) avoid "building" when unnecessary (e.g. tests app)
 2) avoid using GH cache to share charms with tests

The repo has 60 commits, 44 touched tests and 2 only changed test app "application_charm":

```
> mysql-operator.git$ git log --oneline | wc -l
60
> mysql-operator.git$ git log --oneline tests/ | wc -l
44
> mysql-operator.git$ git log --oneline tests/integration/high_availability/application_charm/ | wc -l
2
```

IDEA1: What about to store `PR charms` in `branches` on `<track>/<risk>/<branch>`, e.g. `latest/edge/pr_78_mysql_123456` `latest/edge/pr_78_app1_654321`. So we will use [snapstore](https://snapcraft.io/docs/publish-to-a-branch) and [charmhub](https://juju.is/docs/olm/deploy-a-charm-from-charmhub#heading--specify-a-charmed-operator-channel) as a cache to share charms with tests. 

Sure, we still need GH caching, but only for `building` charms.

IDEA2: Publish test app on charmhub too. As separate application OR branch `latest/edge/app1_rev42`.
Test will use it if `latest/edge/pr_78_app1_654321` is missing.

Logic could be:

* build main charmcraft.yaml with cache (this PR) for all series (focal, jammy) as matrix. Push to `<branch>` for Tests.
* build test app charmcraft.yaml only if some REVISION is bumped manually (2 commits our of 60 in this repo).
* teach tests to use `<track>/<risk>/<branch>` with fallback to `<track>/<risk>` if no charm available in `<branch>`.
* publish charm to `<track>/<risk>` after merging (like now) and clean `<branch>`.

The `branch` could be `pr_$id_$name_$commit`:
  pr_ : some prefix
  $id : GH PR id
  $name : app name: mysql , app1 , app2
  $commit : git commit id (avoid rebuilding the same commit on re-trigger, etc)

What I like here:
* all charms are stored in stores only. And tests are using all charms from store.
* not-yet merged charms stored in branches. Future: PR tests in `mysql-bundle` can use not-yet-merged `mysql-operator` from `<branch>` (through some `topic` label).
* separation of building / using logic => should simplify this PR.

What do you think?

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -10,159 +14,160 @@ jobs:
...
+    outputs:
+      build-matrix: ${{ steps.create-matrix.outputs.build_matrix }}
 
-  integration-test-lxd-db-router:
+  build:
     strategy:
       matrix:
-        series: [focal, jammy]
-    name: Integration tests for db-router relation (lxd)
+        charm: ${{ fromJSON(needs.get-build-matrix.outputs.build-matrix) }}
```
Comment (delgod) [inline .github/workflows/ci.yaml:53]: It looks too complex to use generated matrix here. I would have a matrix explicitly configured here without scripts.
Any reason to have it generated?

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
```
Comment (delgod) [inline tests/integration/conftest.py:1]: just Q: where do we use this file?

**Comment by shayancanonical** [comment]
Thinking of simplifying, I was hoping to discuss the following topics to see if they can help make caching more simple/efficient:

1. from my understanding, `charmcraft pack` reuses LXD containers which need to be cleaned when the following files are modified: metadata.yaml, charmcraft.yaml, requirements.txt and actions.yaml. These files do not change very frequently, so I am wondering if we should compile our cache key using these. when the above files are not changed, the lxc container can be reused to pack any file changes in the charm code.
2. [The following discourse post](https://discourse.charmhub.io/t/charmcraft-bases-provider-support/4713) and the [following lxd documentation](https://linuxcontainers.org/lxd/docs/master/remotes/) indicate that we can set up lxd to use a remote server. I am wondering if using a remote machine with sufficient memory/disk space and a cron to periodically clean lxd projects may make caching simpler

Happy to discuss these offline in more detail (and continuing fleshing out the above topics)

**Comment by jardon** [comment]
Could we at least break apart the caching bits into a separate Github Actions repo that is reusable?  That way we can version control it from a central location and each following PR wont be hundreds of lines of code.

If there's any technical reason this is not possible, lmk.

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        id: charmcraft-version
+        # Major and minor versions (e.g. "2.1")
+        run: echo "version=$(charmcraft version | cut --delimiter '.' --fields 1,2)" >> $GITHUB_OUTPUT
+      - name: Restore cache of `charmcraft pack` LXC instance
+        id: restore-cache
+        uses: actions/cache/restore@v3
+        with:
+          path: ~/ga-charmcraft-cache/**
+          key: charmcraft-pack-${{ matrix.charm.path }}-${{ matrix.charm.bases_index }}-${{ steps.charmcraft-version.outputs.version }}-${{ hashFiles(format('{0}/charmcraft.yaml', matrix.charm.path), format('{0}/actions.yaml', matrix.charm.path), format('{0}/requirements.txt', matrix.charm.path)) }}
+      - name: Import cached container
```
Comment (shayancanonical) [inline .github/workflows/build.yaml:61]: to confirm: if the CI is scheduled (nightly), we will import the cached LXD container for `main`, run charmcraft pack, delete cached LXD container for `main`, and then save the LXD container from the charmcraft pack step?

if so, i have the following suggestion: we should not import any cached LXD containers if scheduled, run charmcraft pack, then save the LXD container from the charmcraft pack step. This will result in a fresh LXD container to be created and cached

**Review by shayancanonical — APPROVED** [review]
LGTM, thanks for iterating!

File: .github/workflows/ci.yaml (hunk 7 above)
Comment (taurus-forever) [reply .github/workflows/ci.yaml]: I think this thread can be resolved. Re-checking all threads to submit this PR.
